runtime/src/bank/check_transactions.rs

```diff
@@ -102,36 +110,52 @@ impl Bank {
                 nonce: None,
                 lamports_per_signature: hash_info.lamports_per_signature(),
             })
-        } else if let Some((nonce, nonce_data)) =
-            self.check_and_load_message_nonce_account(tx.message(), next_durable_nonce)
+        } else if let Some((nonce, lamports_per_signature)) =
+            self.check_load_and_advance_message_nonce_account(tx.message(), next_durable_nonce)
         {
             Ok(CheckedTransactionDetails {
                 nonce: Some(nonce),
-                lamports_per_signature: nonce_data.get_lamports_per_signature(),
+                lamports_per_signature,
             })
         } else {
             error_counters.blockhash_not_found += 1;
             Err(TransactionError::BlockhashNotFound)
         }
     }
 
-    pub(super) fn check_and_load_message_nonce_account(
+    pub(super) fn check_load_and_advance_message_nonce_account(
         &self,
         message: &SanitizedMessage,
         next_durable_nonce: &DurableNonce,
-    ) -> Option<(NonceInfo, nonce::state::Data)> {
+    ) -> Option<(NonceInfo, u64)> {
         let nonce_is_advanceable = message.recent_blockhash() != next_durable_nonce.as_hash();
-        if nonce_is_advanceable {
-            self.load_message_nonce_account(message)
-        } else {
-            None
+        if !nonce_is_advanceable {
+            return None;
         }
+
+        let (nonce_address, mut nonce_account, nonce_data) =
+            self.load_message_nonce_account(message)?;
+
+        let lamports_per_signature = self.get_lamports_per_signature();
```

Comment:
This isn't the correct `lamports_per_signature` value. Instead of getting it from the fee calculator, it needs to come from the blockhash queue. You should grab the value from the `hash_queue` in `check_age`

Reply:
thank you for catching this, i used an unwrap as this is identical to `last_blockhash_and_lamports_per_signature()` but let me know if youd rather i zip and match on an option

btw what is the difference between the `lamports_per_signature` on the bank and the one on the blockhash queue? im not familiar with how variable fees would function if they were ever turned on/employed

Reply:
> btw what is the difference between the `lamports_per_signature` on the bank and the one on the blockhash queue? im not familiar with how variable fees would function if they were ever turned on/employed

In `register_recent_blockhash` we would have recorded the derived lps for the blockhash of the current bank's _**parent block**_. Then in the _current bank's_ constructor (`_new_from_parent`) we derive a _**new**_ lps based on the signature count of the parent block's signature count (`FeeRateGovernor::new_derived`). So it's important that we are using the lps value from the parent bank when populating the nonce, not the value from the current bank.

Now, let's look into how lps could change inside the fee rate governor.. I pulled the fee rate governor fields from a mainnet beta snapshot:
```
[snapshot-tool/src/main.rs:200:13] bank.fee_rate_governor = FeeRateGovernor {
    lamports_per_signature: 5000,
    target_lamports_per_signature: 10000,
    target_signatures_per_slot: 20000,
    min_lamports_per_signature: 5000,
    max_lamports_per_signature: 100000,
    burn_percent: 50,
}
```

Based on the logic for `FeeRateGovernor::new_derived` it looks like a parent bank would need to have over 10,000 signatures in order to increase the desired lps beyond 5000. So this code definitely is enabled but...

Also it's important to call out that for fee calculation we don't really use lps from the `FeeRateGovernor` anymore, we get it from `FeeStructure` (see how `get_fee_for_message_with_lamports_per_signature` only uses lps from fee rate governor to disable fees for tests). We **_only_** use the fee rate governor's lps value for storing in the nonce account right now and that's why it's important to get correct for consensus right now. But we should clean up all uses of the fee rate governor altogether at some point.

Btw, in the past we did have a separate mechanism for congestion driven fees but it was buggy and removed. Discussion here: https://discord.com/channels/428295358100013066/488758828330909706/1201405954256818176